network/src/application/netperf/mod.rs

```diff
@@ -0,0 +1,335 @@
+        network_context: NetworkContext,
+        peers: Arc<PeerMetadataStorage>,
+        sender: Arc<NetPerfNetworkSender>,
+        events: NetPerfNetworkEvents,
+        netperf_port: u16,
+    ) -> Self {
+        NetPerf {
+            network_context,
+            peers,
+            peer_list: Arc::new(DashMap::with_capacity(128)),
+            sender,
+            events,
+            netperf_port,
+        }
+    }
+
+    /// Configuration for the network endpoints to support NetPerf.
+    pub fn network_endpoint_config() -> NetworkApplicationConfig {
+        NetworkApplicationConfig::client_and_service(
+            [
+                ProtocolId::NetPerfDirectSendCompressed,
+                ProtocolId::NetPerfRpcCompressed,
+            ],
+            aptos_channel::Config::new(NETWORK_CHANNEL_SIZE)
+                .queue_style(QueueStyle::FIFO)
+                .counters(&counters::PENDING_NET_PERF_NETWORK_EVENTS),
+        )
+    }
+
+    fn net_perf_state(&self, sender: Sender<NetPerfCommands>) -> NetPerfState {
+        NetPerfState {
+            peers: self.peers.clone(),
+            sender: self.sender.clone(),
+            peer_list: self.peer_list.clone(),
+            tx: sender,
+        }
+    }
+
+    async fn start(mut self) {
+        let port = preferred_axum_port(self.netperf_port);
+        let (tx, _rx) = tokio::sync::mpsc::channel::<NetPerfCommands>(NETPERF_COMMAND_CHANNEL_SIZE);
+
+        info!(
+            NetworkSchema::new(&self.network_context),
+            "{} NetPerf Event Listener started", self.network_context,
+        );
+
+        spawn_named!(
+            "NetPerf Axum",
+            start_axum(self.net_perf_state(tx.clone()), port)
+        );
+
+        loop {
+            futures::select! {
+                maybe_event = self.events.next() => {
+                    // Shutdown the NetPerf when this network instance shuts
+                    // down. This happens when the `PeerManager` drops.
+                    let event = match maybe_event {
+                        Some(event) => event,
+                        None => break,
+                    };
+
+                    match event {
+                        Event::NewPeer(metadata) => {
+                            self.peer_list.insert(
+                                metadata.remote_peer_id,
+                                PeerNetPerfStat::new(metadata)
+                            );
+                        }
+                        Event::LostPeer(metadata) => {
+                            self.peer_list.remove(
+                                &metadata.remote_peer_id
+                            );
+                        }
+                        Event::Message(_peer_id, msg) =>  {
+                            match msg {
+                                NetPerfMsg::BlockOfBytes(_bytes) => {
+                                    /* maybe add dedicated counters? but network_application_{out/in}bound_traffic
```

Comment:
It might make sense to add a NetPerf specific counter here to actually see the throughput of messages making it to the application layer? Even if the network registers the new message and pushes it on the queue, part of me wonders if there's a bottleneck when we bubble it up to the application layer (e.g., if the application isn't pull of the queue fast enough, etc.).